Base/src/main/java/io/deephaven/base/ringbuffer/CharRingBuffer.java

```diff
@@ -0,0 +1,300 @@
+        size = tail = head = 0;
+    }
+
+    /**
+     * Adds an entry to the ring buffer, will throw an exception if buffer is full. For a graceful failure, use
+     * {@link #offer(char)}
+     *
+     * @param e the char to be added to the buffer
+     * @throws UnsupportedOperationException when {@code growable} is {@code false} and buffer is full
+     * @return {@code true} if the char was added successfully
+     */
+    public boolean add(char e) {
+        if (isFull()) {
+            if (!growable) {
+                throw new UnsupportedOperationException("Ring buffer is full and growth is disabled");
+            } else {
+                grow();
+            }
+        }
+        addUnsafe(e);
+        return true;
+    }
+
+    /**
+     * Ensure that there is sufficient empty space to store {@code count} items in the buffer. If the buffer is
+     * {@code growable}, this may result in an internal growth operation. This call should be used in conjunction with
+     * {@link #addUnsafe(char)}.
+     *
+     * @param count the minimum number of empty entries in the buffer after this call
+     * @throws UnsupportedOperationException when {@code growable} is {@code false} and buffer is full
+     */
+    public void ensureRemaining(int count) {
+        if (remaining() < count) {
+            if (!growable) {
+                throw new UnsupportedOperationException("Ring buffer is full and growth is disabled");
+            } else {
+                grow(count);
+            }
+        }
+    }
+
+    /**
+     * Add values unsafely (will silently overwrite values if the buffer is full). This call should be used in
+     * conjunction with {@link #ensureRemaining(int)}.
+     *
+     * @param e the value to add to the buffer
+     */
+    public void addUnsafe(char e) {
+        storage[tail] = e;
+        tail = (tail + 1) % storage.length;
+        size++;
+    }
+
+    /**
+     * Add an entry to the ring buffer. If the buffer is full, will overwrite the oldest entry with the new one.
+     *
+     * @param e the char to be added to the buffer
+     * @param notFullResult value to return is the buffer is not full
+     * @return the overwritten entry if the buffer is full, the provided value otherwise
+     */
+    public char addOverwrite(char e, char notFullResult) {
+        char result = notFullResult;
+        if (isFull()) {
+            result = remove();
+        }
+        addUnsafe(e);
+        return result;
+    }
+
+    /**
+     * Attempt to add an entry to the ring buffer. If the buffer is full, the write will fail and the buffer will not
+     * grow even if allowed.
+     *
+     * @param e the char to be added to the buffer
+     * @return true if the value was added successfully, false otherwise
+     */
+    public boolean offer(char e) {
+        if (isFull()) {
+            return false;
+        }
+        addUnsafe(e);
+        return true;
+    }
+
+    public char[] remove(int count) {
```

Comment:
By using our helper method (still yet to be defined below) and getting rid of the mod operation, and dealing with the problematic integer overflow case (where `head + count` can go negative), this could be rewritten as
```
    public char[] remove(int count) {
        if (size < count) {
            throw new NoSuchElementException();
        }
        final char[] result = new char[count];
        copyRingBufferToArray(result);
        long newHead = (long)head + count;
        if (newHead >= storage.length) {
            newHead -= storage.length;
        }
        head = Math.toIntExact(newHead);
        size -= count;
        return result;
    }
```

Reply:
These were all great comments and suggestions. Thanks for a thorough review!